### PR TITLE
screen transitions for mercury UI

### PR DESCRIPTION
**Pull request by TychoVrahe**
screen transitions for mercury UI

### CODE REVIEW

**Comment by matejcik** [comment]
these two commits save 2.5 kB + alignment:
f37200210fcaf2736169c714210f9820b25aef77 is essentially @mmilata 's suggestion. uncontroversial, please cherry pick (or do your own version
4da54ecd92b4708e2a59451b55f741e1f442bc6c eliminates a lot of repetition out of the swipe detect code, but needs more careful review because i'm not sure if i didn't mess up some pre-existing functionality

File: core/embed/rust/src/ui/model_mercury/flow/get_address.rs [hunk 1]
```diff
@@ -167,30 +175,36 @@ impl GetAddress {
...
-        let content_address = Frame::left_aligned(title, SwipePage::vertical(paragraphs))
-            .with_menu_button()
-            .with_footer(TR::instructions__swipe_up.into(), None)
-            .map(|msg| matches!(msg, FrameMsg::Button(_)).then_some(FlowMsg::Info))
-            .one_button_request(ButtonRequest::from_num(br_code, br_type))
-            // Count tap-to-confirm screen towards page count
-            .with_pages(|address_pages| address_pages + 1);
+        let content_address = Frame::left_aligned(
+            title,
+            SwipeContent::new(SwipePage::vertical(paragraphs.clone())),
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/model_mercury/flow/get_address.rs]: https://github.com/trezor/trezor-firmware/commit/9854b9adf2af1231531a68e43300045829b5ea58

File: core/embed/rust/src/ui/flow/base.rs [hunk 2]
```diff
@@ -25,6 +25,12 @@ pub trait Swipable<T> {
     }
 }
 
+pub trait SimpleSwipable {
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/flow/base.rs]: fixed https://github.com/trezor/trezor-firmware/commit/dfd4fde1ebd097ce938035264f516b20b681ee04

File: core/embed/rust/src/ui/flow/page.rs [hunk 3]
```diff
@@ -28,7 +26,6 @@ pub struct SwipePage<T> {
     axis: Axis,
     pages: usize,
     current: usize,
-    transition: Option<Transition<T>>,
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/flow/page.rs:31]: removed https://github.com/trezor/trezor-firmware/commit/bbb40ce9a08e783c369ad1d75fd1030f2858772b

File: core/embed/rust/src/ui/flow/store.rs [hunk 4]
```diff
@@ -149,28 +159,40 @@ where
...
+        }
+    }
+
+    fn get_internal_page_count(&mut self, i: usize) -> usize {
+        if i == 0 {
+            self.as_mut().get_internal_page_count()
+        } else {
+            self.next.get_internal_page_count(i - 1)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/flow/store.rs:189]: fixed https://github.com/trezor/trezor-firmware/commit/bbb40ce9a08e783c369ad1d75fd1030f2858772b

File: core/embed/rust/src/ui/component/base.rs [hunk 5]
```diff
@@ -466,6 +468,20 @@ where
...
+    #[cfg(feature = "touch")]
+    Swipe(SwipeDirection),
+}
+
+#[cfg(feature = "touch")]
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum SwipeEvent {
+    Move(SwipeDirection, i16),
+    End(SwipeDirection),
+}
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/component/base.rs]: moved https://github.com/trezor/trezor-firmware/commit/bbb40ce9a08e783c369ad1d75fd1030f2858772b

File: core/embed/rust/src/ui/component/base.rs [hunk 6]
```diff
@@ -466,6 +468,20 @@ where
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum AttachType {
+    Normal,
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/component/base.rs]: initial it is https://github.com/trezor/trezor-firmware/commit/594424138161a170599f92d7e958d5032cb7feaa

File: core/embed/rust/src/ui/flow/base.rs [hunk 7]
```diff
@@ -25,6 +25,12 @@ pub trait Swipable<T> {
...
+    fn get_swipe_config(&self) -> SwipeConfig;
+
+    fn get_internal_page_count(&mut self) -> usize;
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/flow/base.rs]: fixed https://github.com/trezor/trezor-firmware/commit/a303daa8ff0c79aa611e2f297ecd282c3df1411d

File: core/embed/rust/src/ui/model_mercury/component/share_words.rs [hunk 8]
```diff
@@ -160,48 +198,48 @@ impl<'a> Component for ShareWords<'a> {
...
+//         // ));
+//         // ctx.request_anim_frame();
+//         // ctx.request_paint();
+//         // SwipableResult::Animating
+//     }
+//
+//     fn swipe_finished(&self) -> bool {
+//         self.animation.is_none()
+//     }
+//}
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/model_mercury/component/share_words.rs]: https://github.com/trezor/trezor-firmware/commit/dfd4fde1ebd097ce938035264f516b20b681ee04

File: core/embed/rust/src/ui/component/swipe_detect.rs [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+    }
+
+    pub const fn immediate() -> Self {
+        Self {
+            duration: Duration::from_millis(0),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/rust/src/ui/component/swipe_detect.rs]: done by matejcik in https://github.com/trezor/trezor-firmware/commit/888e29725083d422604e854cc3c8407f2c0ddf78

**Comment by TychoVrahe** [comment]
@mmilata @matejcik 

Cherry-picked your commits. 

@matejcik 's refactoring of the SwipeDetect had some bugs, see https://github.com/trezor/trezor-firmware/commit/eedd5d7029982e1fd746c51d1683d3bc86793645 , otherwise i think is nice improvement

File: core/embed/rust/src/ui/component/swipe_detect.rs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+                            final_value,
+                            Duration::from_millis(duration),
+                            Instant::now(),
+                        ));
+                    } else {
+                        // clear animation
+                        self.final_animation = None;
+                        self.moved = 0;
+                        self.locked = None;
+                        return Some(SwipeDetectMsg::Trigger(locked));
```
Comment (matejcik) [inline core/embed/rust/src/ui/component/swipe_detect.rs:321]: I think this is a bug now? When doing the refactor I didn't understand that you can animate back without triggering.

Now when animations are disabled, and you go back (don't finish the locked swipe), you will still trigger. We'd need to do `if final_value > 0` here, right?
(question is, do we even care?)

**Review by matejcik — APPROVED** [review]
leaving an approve because other than my last comment i'm OK with this

**Review by mmilata — APPROVED** [review]

**Comment by STew790** [comment]
**QA OK**

**Info:**
 - OS: Linux x86_64
 - Screen: 1920x1080
 - Device: Trezor T3T1 2.7.2 regular (revision da75d8f4b67410b40a9cfd2954d183d81dd6e8e8)
 - Transport: BridgeTransport 2.0.33

**Pull request merged**